clarity/src/vm/analysis/type_checker/tests/contracts.rs

```diff
@@ -630,3 +664,1864 @@ fn test_expects() {
+    type_check(&contract_id, &mut contract, db, true).map_err(|e| e.to_string())
+}
+
+fn call(
+    db: &mut AnalysisDatabase,
+    contract: &str,
+    function: &str,
+    args: &str,
+) -> Result<ContractAnalysis, String> {
+    let source = format!("(contract-call? .{} {} {})", contract, function, args);
+    let contract_id = QualifiedContractIdentifier::transient();
+    let mut contract = parse(
+        &contract_id,
+        source.as_str(),
+        ClarityVersion::latest(),
+        StacksEpochId::latest(),
+    )
+    .map_err(|e| e.to_string())?;
+    type_check(&contract_id, &mut contract, db, false).map_err(|e| e.to_string())
+}
+
+#[test]
+fn clarity_trait_experiments_impl() {
+    let version = ClarityVersion::latest();
+    let epoch = StacksEpochId::latest();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    let result = db.execute(|db| {
+        load(db, "math-trait")?;
+        load(db, "impl-math-trait")
+    });
+    match result {
+        Ok(_) => (),
+        res => panic!("expected success, got {:?}", res),
+    };
+}
+
+#[test]
+fn clarity_trait_experiments_use() {
+    let version = ClarityVersion::latest();
+    let epoch = StacksEpochId::latest();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    let result = db.execute(|db| {
+        load(db, "math-trait")?;
+        load(db, "use-math-trait")
+    });
+    match result {
+        Ok(_) => (),
+        res => panic!("expected success, got {:?}", res),
+    };
+}
+
+#[test]
+fn clarity_trait_experiments_empty_trait() {
+    let version = ClarityVersion::latest();
+    let epoch = StacksEpochId::latest();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    // Can we define an empty trait?
+    let result = db.execute(|db| load(db, "empty-trait"));
+    match result {
+        Ok(_) => (),
+        res => panic!("expected success, got {:?}", res),
+    };
+}
+
+#[test]
+fn clarity_trait_experiments_duplicate_trait() {
+    let version = ClarityVersion::latest();
+    let epoch = StacksEpochId::latest();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    // Can we re-define a trait with the same type and same name in a different contract?
+    let result = db.execute(|db| {
+        load(db, "empty-trait")?;
+        load(db, "empty-trait-copy")
+    });
+    match result {
+        Ok(_) => (),
+        res => panic!("expected success, got {:?}", res),
+    };
+}
+
+#[test]
+fn clarity_trait_experiments_use_undefined() {
+    let version = ClarityVersion::latest();
+    let epoch = StacksEpochId::latest();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    // Can we define traits that use traits in not-yet-deployed contracts?
+    let result = db.execute(|db| load(db, "no-trait"));
+    match result {
+        Err(_) => (),
```

Comment:
Do you want to test the specific error (or at least the category of error)?

Reply:
Error checks added in d247e4e4b17cbc4c0f43df638cf2af6a3d31ff94